rllib/models/base_model.py

```diff
@@ -0,0 +1,244 @@
+
+
+class RecurrentModel(abc.ABC):
+    """The base model all other models are based on."""
+
+    def __init__(self, name: Optional[str] = None):
+        self._name = name or self.__class__.__name__
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    # TODO: Should these be properties or functions? How do we know the
+    # input_spec before init? Won't our model change shapes depending
+    # on arguments passed to init?
+    @property
+    @abc.abstractmethod
+    def input_spec(self) -> SpecDict:
+        """Returns the spec of the input of this module."""
+
+    @property
+    @abc.abstractmethod
+    def prev_state_spec(self) -> SpecDict:
+        """Returns the spec of the prev_state of this module."""
+
+    @property
+    @abc.abstractmethod
+    def output_spec(self) -> SpecDict:
+        """Returns the spec of the output of this module."""
+
+    @property
+    @abc.abstractmethod
+    def next_state_spec(self) -> SpecDict:
+        """Returns the spec of the next_state of this module."""
+
+    @abc.abstractmethod
+    def _initial_state(self) -> TensorDict:
+        """Initial state of the component.
+        If this component returns a next_state in its unroll function, then
+        this function provides the initial state.
+        Subclasses should override this function instead of `initial_state`, which
+        adds additional checks.
+
+        Returns:
+          A TensorDict containing the state before the first step.
+        """
+
+    def initial_state(self) -> TensorDict:
+        """Initial state of the component.
+        If this component returns a next_state in its unroll function, then
+        this function provides the initial state.
+
+        Returns:
+          A TensorDict containing the state before the first step.
+        """
+        initial_state = self._initial_state()
+        self.next_state_spec.validate(initial_state)
+        return initial_state
+
+    @abc.abstractmethod
+    def _unroll(
+        self, inputs: TensorDict, prev_state: TensorDict, **kwargs
+    ) -> UnrollOutputType:
+        """Computes the output of the module over unroll_len timesteps.
+        Subclasses should override this function instead of `unroll`, which
+        adds additional checks.
+
+        Args:
+          inputs: A TensorDict of inputs
+          prev_state: A TensorDict containing the next_state of the last
+            timestep of the previous unroll.
+        Returns:
+          outputs: A TensorDict of outputs
+          next_state: A dict containing the state to be passed
+            as the first state of the next rollout.
+        """
+
+    def unroll(
+        self, inputs: TensorDict, prev_state: TensorDict, **kwargs
+    ) -> UnrollOutputType:
+        """Computes the output of the module over unroll_len timesteps.
+
+        Args:
+          inputs: A TensorDict containing inputs to the model
+          prev_state: A TensorDict containing containing the
```

Comment:
can we add here that kwargs is for forward compatibility here? Also can we add this to `init_state()`